### PR TITLE
`azurerm_storage_account` - add supports of `{table|queue}_encryption_key_type`

### LINKED ISSUE #13999: Support for CMK-enabled Queue Storage Accounts
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

On azurerm Version 2.83.0, there is currently no support for creating Azure storage accounts with customer-managed key encryption of table and queue data. This requires the ability to change encryption key-type from "Service" to "Account" for the two data types.

As per Azure queue storage documentation [here](https://docs.microsoft.com/en-gb/azure/storage/common/account-encryption-key-create?tabs=template), at the point of storage account creation the 'account' encryption key must be set for queues and tables. Otherwise, only Blob storage and Azure Files will use the account encryption key to encrypt data.

The above azure documentation lists how this is achieved with the Azure Portal / CLI / ARM. However, terraform does not seem to have support for this and therefore accounts are created with 'blobs and files only' CMK-encryption.

image

Our use case requires CMK-encryption of storage queues. Using terraform is highly preferred and we only want to look at other options as a last resort, so hopefully this can be resolved soon.

Suggested fix would be like the[ azure CLI options](https://docs.microsoft.com/en-us/cli/azure/storage/account?view=azure-cli-latest#az_storage_account_create). Adding `encryption-key-type-for-queue` and `encryption-key-type-for-table` arguments onto azurerm_storage_account.

### New or Affected Resource(s)

* azurerm_storage_account - resource and data source
* azurerm_storage_account_customer_managed_key - no changes required.

### Potential Terraform Configuration

```hcl
resource "azurerm_storage_account" "example" {   
  name                     = "storageaccountname"   
  resource_group_name      = azurerm_resource_group.example.name   
  location                 = azurerm_resource_group.example.location   
  account_tier             = "Standard"   
  account_replication_type = "GZRS"     
  encryption-key-type-for-queue = "Account"   
  encryption-key-type-for-table = "Account" 
}
```
This copies the behaviour for [azure CLI](https://docs.microsoft.com/en-us/cli/azure/storage/account?view=azure-cli-latest#az_storage_account_create).

### Design

Update the terraform [azurerm_storage_account](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/storage/storage_account_resource.go) resource and data source to take two new config options:

- encryption-key-type-for-queue
- encryption-key-type-for-table

#### Changes to the terraform code
Update the storage_account_resource code for the CRUD operations as below.

**CREATE operations:**

1. Add the encryption-key-type-for-queue /table options to the schema for this resource [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/storage/storage_account_resource.go#L65)

2. Add an 'Encryption' field to the [AccountPropertiesCreateParameters](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go@v54.3.0+incompatible/services/storage/mgmt/2021-01-01/storage#AccountPropertiesCreateParameters) instantiation for the storage_account_resource[ here](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/storage/storage_account_resource.go#L934)

This will use the [Encyrption](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go@v54.3.0+incompatible/services/storage/mgmt/2021-01-01/storage#Encryption) field already present on the Azure Go SDK.

pseudocode:
```
If  encryption-key-type-for-queue == "Account":
   set encryption = {
        "services": {
          "queue": {
             "keyType": "Account"
           }
        }
If encryption-key-type-for-table = "Account":
   set encryption = {
        "services": {
          "table":{
             "keyType": "Account"
           }
        }

parameters := storage.AccountCreateParameters{
		Location: &location,
		Sku: &storage.Sku{
			Name: storage.SkuName(storageType),
		},
		Tags: tags.Expand(t),

		Kind: storage.Kind(accountKind),
		AccountPropertiesCreateParameters: &storage.AccountPropertiesCreateParameters{
			EnableHTTPSTrafficOnly: &enableHTTPSTrafficOnly,
			NetworkRuleSet:         expandStorageAccountNetworkRules(d, tenantId),
			IsHnsEnabled:           &isHnsEnabled,
                        EnableNfsV3:            &nfsV3Enabled,
			AllowSharedKeyAccess:   &allowSharedKeyAccess,
                        Encryption: &encryption,
		},
	}
```
Likewise if both options are given then set both to "Account".

**Validation of the above design:**

The above has been tested with Azure REST API (which the Azure Go SDK is based on), this will enable CMK encryption and change keytype to "Account" for the chosen service. Blob and file remain unaffected because they're always set.

Request body:
```
{
    "sku":{
       "name": "Standard_GZRS"
     },
    "kind": "StorageV2",
    "location": "uksouth",
    "identity":{
         "type": "SystemAssigned"
     },
    "properties": {
      "allowBlobPublicAccess": false,
      "minimumTlsVersion": "TLS1_2",
      "allowSharedKeyAccess": true,
      "encryption": {
        "services": {
          "queue":{
             "keyType": "Account"
           },
          "table":{
             "keyType": "Account"
           }
        },
        "keySource": "Microsoft.Storage"
      }
    }
}
```
 Returns config:
```
Response content:
{
….
    "encryption": {
      "keySource": "Microsoft.Storage",
      "services": {
        "blob": {
          "enabled": true,
          "keyType": "Account",
          "lastEnabledTime": "2021-11-04T14:49:46.5849143Z"
        },
        "file": {
          "enabled": true,
          "keyType": "Account",
          "lastEnabledTime": "2021-11-04T14:49:46.5849143Z"
        },
        "queue": {
          "enabled": true,
          "keyType": "Account",
          "lastEnabledTime": "2021-11-04T14:49:46.5849143Z"
        },
        "table": {
          "enabled": true,
          "keyType": "Account",
          "lastEnabledTime": "2021-11-04T14:49:46.5849143Z"
        }
      }
    },
…
}
```
Without these fields, the queue and table are not CMK enabled.

Following this I've then tested adding CMKs to the account. The current terraform implementation of azurerm_storage_account_cmk works as designed so I don't think we need to make changes there.

**UPDATE**
On azure rest API we cannot change the keytypes of services after the initial creation of the storage account. 

e.g. Changing queue/table from "Service" to "Account" and visa versa returns the following error
```
{"error":{"code":"AccountPropertyCannotBeUpdated","message":"The property 'keyType' was specified in the input, but it cannot be updated as it is read-only. For more information, see - https://aka.ms/storageaccountupdate"}}
```
Therefore we should add checks for this too. Of the form:
```
If  d.HasChange(encryption-key-type-for-queue): 
    return fmt.Errorf(`"encryption-key-type-for-queue" cannot be changed as it is read-only. Current configured value: %s, Attempted replacement: %s"}})


If d.HasChange(encryption-key-type-for-table):
    return fmt.Errorf(`"encryption-key-type-for-queue" cannot be changed as it is read-only. Current configured value: %s, Attempted replacement: %s"}})
```

**READ**
`resourceStorageAccountRead` returns the `encryption-key-type-for-queue` / `-table` fields + values
```
d.set("encryption-key-type-for-queue", props.encryption.services.queue.KeyType)
d.set("encryption-key-type-for-table", props.encryption.services.table.KeyType)
```

**DELETE**
No changes required.
 
**Changes to the storage_account_data_source**

Like the READ operation above - data source returns the encryption-key-type-for-queue / table fields + values
```
d.set("encryption-key-type-for-queue", props.encryption.services.queue.KeyType)
d.set("encryption-key-type-for-table", props.encryption.services.table.KeyType)
```
Update docs and tests in line with this^

 

Testing in azurerm terraform code:

Add tests for this in azurerm_storage_account_resource_test file 

Docs Updates:

Update the terraform docs to explain the new fields, as [here](https://github.com/hashicorp/terraform-provider-azurerm/pull/13014/commits/02a58f985e81af36bdf5bd8fdea5845d685b316a#diff-9b6a86c165940b9332b262be7ef74223941f02a9c05ffe5876aad84a3bd1d5fd)

### Acceptance Criteria

- Design / implementation Reviewed by the azure terraform
- Design / implementation Review by the terraform storage account devs
- We can successfully use terraform create accounts with encryption for azure queue storage, with TODO markers for table options.
- Merged to terraform master, minor version bumped, brought into MVM
- Sufficient set of tests passing in terraform code:
- Most of the following tests are probably required before merging to terraform master, we probably can't pull a beta resource provider into simpl-prerequisites to test that.
- That we can successfully create the account with CMKs for blob storage. Is there a way to read the resource that has been created, e.g. azure cli calls to pull config and read settings.
- Test that the config settings pass through to azure rm config successfully.
- Test that CMK creation on queue storage works. Test that queues are encrypted with CMKs?
- Test that private endpoint resource can be created on the account.
- Test that queue resource can be created on the account
- Terraform documentation updated to explain the new field.

### References

https://docs.microsoft.com/en-gb/azure/storage/common/account-encryption-key-create

https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/storage/storage_account_resource.go

* #0000

### PR DESCRIPTION
**Pull request by magodo**
Fixes #13999 (the issue has all the details)

## Test

```shell                                           
💤 TF_ACC=1 go test -timeout=3h -v ./internal/services/storage -run='TestAccStorageAccount_encryptionKeyType'
=== RUN   TestAccStorageAccount_encryptionKeyType
=== PAUSE TestAccStorageAccount_encryptionKeyType                                                                                       
=== CONT  TestAccStorageAccount_encryptionKeyType                                                                                                                                                                                                                               
--- PASS: TestAccStorageAccount_encryptionKeyType (174.58s)                                                                             
PASS                                                                                                                                    
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/storage       174.605s
💤 TF_ACC=1 go test -timeout=3h -v ./internal/services/storage -run='TestAccDataSourceStorageAccount_withEncryptionKey'
=== RUN   TestAccDataSourceStorageAccount_withEncryptionKey
=== PAUSE TestAccDataSourceStorageAccount_withEncryptionKey
=== CONT  TestAccDataSourceStorageAccount_withEncryptionKey
--- PASS: TestAccDataSourceStorageAccount_withEncryptionKey (258.98s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/storage       259.009s
```

### CODE REVIEW

**Comment by sambyng-msft** [comment]
Thanks for implementing this! 

Looks good. For my knowledge, why have you not updated the following?

- storage_account_resource: UPDATE operations with a check for whether the field has changed.
- storage_account_data_source to return the new fields.

[As per the design i wrote last friday. https://github.com/hashicorp/terraform-provider-azurerm/issues/13999]

Aside from the above, can we really try to get this in for the release on Thursday this week (the 11th November)? 

My team and I need this feature as soon as possible - thanks!

**Comment by magodo** [comment]
@sambyng-msft Thanks for pointing this out! I've added a new commit to include the datasource. For the update, since these properties are `ForceNew`, which means when users try to modify them, the TF will warn the users that the next apply will replace the current resource. That said, the update of these two properties is eliminated by TF itself.

**Comment by sambyng-msft** [comment]
> That said, the update of these two properties is eliminated by TF itself.

Can you clarify - wouldn't this introduce a bug? 
e.g. User attempts to update the field from "Service" to "Account", expecting a new resource to be created with "Account", but then TF 'eliminates' the properties so the default "Service" is given. That's not the expected behaviour. Would we be better without any TF elimination at all or have I misunderstood?

**Comment by lonegunmanb** [comment]
> > That said, the update of these two properties is eliminated by TF itself.
> 
> Can you clarify - wouldn't this introduce a bug? e.g. User attempts to update the field from "Service" to "Account", expecting a new resource to be created with "Account", but then TF 'eliminates' the properties so the default "Service" is given. That's not the expected behaviour. Would we be better without any TF elimination at all or have I misunderstood?

Hi @sambyng-msft , what @magodo means I think is, in Terraform, if a property is marked as `ForceNew`, when user modify this property, Terraform would destroy the old resource and create a new one to replace it, with latest config that user defined via Terraform code. It often happened when user want to update a property that has no way to update via API and SDK (for example the property is part of resource id).

Since there's no way to update some properties via API, and Terraform GUARANTEE the resource config in real world would be same as user defined in code, a `ForceNew` would be required to be declared on these properties.

As in this case, I think it depends on whether this property updatable via API, otherwise `ForceNew` is the best we could do for now.

**Comment by sambyng-msft** [comment]
@lonegunmanb That makes sense now, thanks for clarifying. When `ForceNew` is present the ability to update these properties has effectively been 'eliminated'. 

Looks good then. 

@mbfrahry @katbyte Is there any chance that we could get this merged before Thursday?

File: website/docs/r/storage_account.html.markdown [hunk 1]
```diff
@@ -139,6 +139,12 @@ The following arguments are supported:
 
 * `routing` - (Optional) A `routing` block as defined below.
 
+* `queue_encryption_key_type` - (Optional) The encryption type of the queue service. Possible values are `Service` and `Account`. Default value is `Service`.
+
+* `table_encryption_key_type` - (Optional) The encryption type of the table service. Possible values are `Service` and `Account`. Default value is `Service`.
+
```
Comment (aristosvo) [inline website/docs/r/storage_account.html.markdown:145]:
```suggestion
* `queue_encryption_key_type` - (Optional) The encryption type of the queue service. Possible values are `Service` and `Account`. Default value is `Service`. Changing this forces a new resource to be created.

* `table_encryption_key_type` - (Optional) The encryption type of the table service. Possible values are `Service` and `Account`. Default value is `Service`. Changing this forces a new resource to be created.

```

**Review by aristosvo — COMMENTED** [review]
Just a minor comment on the docs, as changing `{table|queue}_encryption_key_type` forces recreation.

**Comment by sambyng-msft** [comment]
@magodo It looks like update to 2021-04-01 storage account requires 'storageV2' to change to 'KindStorageV2'

https://pkg.go.dev/github.com/Azure/azure-sdk-for-go@v54.3.0+incompatible/services/storage/mgmt/2021-01-01/storage#Account
to
https://pkg.go.dev/github.com/Azure/azure-sdk-for-go@v54.3.0+incompatible/services/storage/mgmt/2021-04-01/storage#Account

**Review by aristosvo — APPROVED** [review]
Looks great, thanks!

@sambyng-msft I cannot help with merging, just a community maintainer 🤷🏽‍♂️

File: internal/services/storage/storage_account_resource.go [hunk 2]
```diff
@@ -1678,6 +1726,23 @@ func resourceStorageAccountRead(d *pluginsdk.ResourceData, meta interface{}) err
 			allowSharedKeyAccess = *props.AllowSharedKeyAccess
 		}
 		d.Set("shared_access_key_enabled", allowSharedKeyAccess)
+
+		// Setting the encryption key type to "Service" in PUT. The following GET will not return the queue/table in the service list of its response.
+		// So defaults to setting the encryption key type to "Service" if it is absent in the GET response. Also, define the default value as "Service" in the schema.
+		var (
+			queueEncryptionKeyType = "Service"
+			tableEncryptionKeyType = "Service"
```
Comment (tombuildsstuff) [inline internal/services/storage/storage_account_resource.go:1734]: can we use the constants here

```suggestion
			queueEncryptionKeyType = string(storage.KeyTypeService)
			tableEncryptionKeyType = string(storage.KeyTypeService)
```

File: internal/services/storage/storage_account_data_source.go [hunk 3]
```diff
@@ -368,6 +378,23 @@ func dataSourceStorageAccountRead(d *pluginsdk.ResourceData, meta interface{}) e
 			}
 			d.Set("secondary_blob_connection_string", secondaryBlobConnectStr)
 		}
+
+		// Setting the encryption key type to "Service" in PUT. The following GET will not return the queue/table in the service list of its response.
+		// So defaults to setting the encryption key type to "Service" if it is absent in the GET response. Also, define the default value as "Service" in the schema.
+		var (
+			queueEncryptionKeyType = "Service"
+			tableEncryptionKeyType = "Service"
```
Comment (tombuildsstuff) [inline internal/services/storage/storage_account_data_source.go:386]: can we use the constants here:

```suggestion
			queueEncryptionKeyType = string(storage.KeyTypeService)
			tableEncryptionKeyType = string(storage.KeyTypeService)
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @magodo 

Thanks for this PR - I've taken a look through and left a couple of comments inline but if we can fix those up then this should otherwise be good to merge 👍 

Thanks!

**Review by stephybun — APPROVED** [review]
Thanks for this PR @magodo. Apologies for the delay and LGTM! 👍🏼

**Comment by sambyng-msft** [comment]
thanks!

**Pull request merged**